presto-main/src/main/java/io/prestosql/operator/unnest/BlockProducerFromSource.java

```diff
@@ -0,0 +1,337 @@
+        ids = new int[0];
+        positionCount = 0;
+        usingCopiedBlock = false;
+
+        // clear copied block state
+        outputBlockBuilder = null;
+        pageBuilderStatus = null;
+    }
+
+    /**
+     * Prepares for a new output block. The caller has to ensure that
+     * the source is not null when this method is invoked.
+     * @param pageBuilderStatus
+     * @param expectedEntries
+     */
+    public void startNewOutput(PageBuilderStatus pageBuilderStatus, int expectedEntries)
+    {
+        checkState(source != null, "source is null");
+        this.pageBuilderStatus = pageBuilderStatus;
+
+        // Prepare for a new dictionary block
+        this.ids = new int[expectedEntries];
+        this.positionCount = 0;
+        this.usingCopiedBlock = false;
+    }
+
+    public Block buildOutputAndFlush()
+    {
+        Block outputBlock;
+        if (usingCopiedBlock) {
+            outputBlock = outputBlockBuilder.build();
+        }
+        else {
+            outputBlock = new DictionaryBlock(positionCount, source, ids);
+        }
+
+        // Flush stored state, so that ids can not be modified after
+        // the shallow copy has been used for constructing the dictionary
+        clearCurrentOutput();
+
+        return outputBlock;
+    }
+
+    /**
+     * Prepare for appending element at position {@code index} in the source block
+     * to output block
+     */
+    public void appendElementFromSource(int index)
+    {
+        if (usingCopiedBlock) {
+            type.appendTo(source, index, outputBlockBuilder);
+        }
+        else {
+            checkIndexValid(index);
+            appendId(index);
+        }
+    }
+
+    /**
+     * Prepare for appending range of elements starting at {@code startPosition} and
+     * length {@code length} from the source block to output block.
+     *
+     * Purpose of this method is to avoid repeated range checks for every
+     * invocation of {@link #appendElementFromSource}.
+     * @param startPosition
+     * @param length
+     */
+    public void appendRange(int startPosition, int length)
+    {
+        checkRangeValid(startPosition, length);
+
+        if (usingCopiedBlock) {
+            for (int i = 0; i < length; i++) {
+                type.appendTo(source, startPosition + i, outputBlockBuilder);
+            }
+        }
+        else {
+            for (int i = 0; i < length; i++) {
+                appendId(startPosition + i);
+            }
+        }
+    }
+
+    /**
+     * Prepare for appending the element at position {@code index} in the source
+     * for {@code count} number of times in the output
+     *
+     * Purpose of this method is to avoid repeated checks (for the same
+     * element) in {@link #appendElementFromSource}.
+     * @param index
+     * @param count
+     */
+    public void appendRepeated(int index, int count)
+    {
+        checkIndexValid(index);
+
+        if (usingCopiedBlock) {
```

Comment:
Is this code path possible? I think not:
* If you're using a copied block, then you injected a null
* A null is only needed for misalignment of unnested columns
* appendRepeated is only useful for replicated columns

It's good to have the check here and might be good to have it pre-implemented for some future need, but I'm thinking that this might hint that we're complicating things by trying to handle two different use cases in the same class: replicating positions of a block  (for replicated columns) and inserting nulls into a block (for unnested columns).

Keeping them together in one class may be marginally fewer lines of code, but I think it's more complex and will be harder to add special optimizations to in the future.